File: Python/algorithmic_heights/20_par3.py
# 3-Way Partition
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def three_way_partitioning(array, pivot, n):
    middle_ix = 0
    end_ix = n - 1
    current_ix = 1
    while current_ix <= end_ix:
        elem = array[current_ix]
        if elem < pivot:
            array[current_ix], array[middle_ix] = array[middle_ix], elem
            current_ix += 1
            middle_ix += 1
        elif elem > pivot:
            array[current_ix], array[end_ix] = array[end_ix], elem
            end_ix -= 1
        else:
            current_ix += 1


def main():
    reader = sys.stdin
    n = int(next(reader))
    array = convert_to_intlist(next(reader))
    three_way_partitioning(array, array[0], n)
    print(' '.join(map(str, array)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/20_par3.hs
import Control.Monad.ST (runST)
import Data.Vector (Vector, freeze, fromList, thaw, toList)
import qualified Data.Vector as V
import qualified Data.Vector.Mutable as MV

convertToIntList :: String -> [Int]
convertToIntList = map read . words


threeWayPartitioning :: Vector Int -> Int -> Int -> Vector Int
threeWayPartitioning vector pivot n =
    let swapElements array =
            let go middleIx currentIx endIx =
                    if currentIx > endIx then return ()
                    else do
                        elem <- MV.read array currentIx
                        if elem < pivot then do
                            MV.swap array currentIx middleIx
                            go (middleIx + 1) (currentIx + 1) endIx
                        else if elem > pivot then do
                            MV.swap array currentIx endIx
                            go middleIx currentIx (endIx - 1)
                        else go middleIx (currentIx + 1) endIx
            in go 0 1 (n - 1)
    in runST $ do
        array <- thaw vector
        swapElements array
        freeze array


main :: IO ()
main = do
    n <- readLn
    xs <- fromList . convertToIntList <$> getLine
    let result = threeWayPartitioning xs (V.head xs) n
    putStrLn $ unwords $ map show $ toList result
